### PR TITLE
Feat: Live-1 Ingress Report

### PR DESCRIPTION
**Pull request by jackstockley89**
This report will collect information to the cluster for anything using the live-1 ingress dns address and list the Namespace, Ingress name and Host that is using the old address

### CODE REVIEW

File: reports/live-1-ingress/main.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+// Live1DoaminSearch searches all ingress resources and returns a list of namespace, ingress resources and the hosts that are still using the live1-domain name.
+// if there is one. Due to the requirement of the API, we have to sculpt the []byte data a very specific way.
+func Live1IngressSearch(ingressList *v1beta1.IngressList) ([]map[string]string, error) {
+	// s contains a slice of maps, each map will be iterated over when placed in a dashboard.
+	s := make([]map[string]string, 0)
+	for _, i := range ingressList.Items {
+
+		for _, v := range i.Spec.TLS {
+			fmt.Println("TLS Loop")
```
Comment (poornima-krishnasamy) [inline reports/live-1-ingress/main.go:90]: can u remove the print statement

**Review by poornima-krishnasamy — COMMENTED** [review]
@jackstockley89 , this looks good. Is it possible to change the dir name to something like `live-1-domains`

File: reports/live-1-ingress/main.go (hunk 1 above)
Comment (jackstockley89) [reply reports/live-1-ingress/main.go]: I have renamed the directory and the function, also removed the line print

File: reports/live-1-domains/main.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	// Get all ingress resources
+	ingressList, err := ingress.GetAllIngressesFromCluster(clientset)
+	if err != nil {
+		log.Fatalln(err.Error())
+	}
+
+	// Find all ingress resources without the required external-dns annotations
+	ingressesLive1DoaminSearch, err := Live1DomainSearch(ingressList)
```
Comment (jasonBirchall) [inline reports/live-1-domains/main.go:55]: Can you rename this variable - perhaps `domainSearch`?

File: reports/live-1-domains/main.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+	// Post json to hoowdaw api
+	err = hoodaw.PostToApi(jsonToPost, hoodawApiKey, &endPoint)
+	if err != nil {
+		log.Fatalln(err.Error())
+	}
+}
+
+// GetAllIngresses takes a Kubernetes clientset and returns all ingress with type *v1beta1.IngressList and an error.
+func GetAllIngresses(clientset *kubernetes.Clientset) (*v1beta1.IngressList, error) {
```
Comment (jasonBirchall) [inline reports/live-1-domains/main.go:73]: It doesn't look like this function is used.

File: reports/live-1-domains/main.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	ingressList, err := clientset.NetworkingV1beta1().Ingresses("").List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return ingressList, nil
+}
+
+// Live1DomainSearch searches all ingress resources and returns a list of namespace, ingress resources and the hosts that are still using the live1-domain name.
+// if there is one. Due to the requirement of the API, we have to sculpt the []byte data a very specific way.
+func Live1DomainSearch(ingressList *v1beta1.IngressList) ([]map[string]string, error) {
```
Comment (jasonBirchall) [inline reports/live-1-domains/main.go:83]: I can't see why this is an exported (public) function. I might be missing something...

File: reports/live-1-domains/main.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+				m["hostname"] = v.Hosts[0]
+				s = append(s, m)
+			}
+		}
+	}
+	return s, nil
+}
+
+// BuildJsonMap takes a slice of maps and return a json encoded map
+func BuildJsonMap(ingressesLive1DoaminSearch []map[string]string) ([]byte, error) {
```
Comment (jasonBirchall) [inline reports/live-1-domains/main.go:102]: Same as above RE public functions.

Comment (jasonBirchall) [reply reports/live-1-domains/main.go:102]: in fact, hasn't this been re-written somewhere. Should we reuse the current function? I'll see if I can dig it out

File: reports/live-1-domains/main.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	endPoint      = *hoodawHost + *hoodawEndpoint
+)
+
+func main() {
+	flag.Parse()
+
+	// Gain access to a Kubernetes cluster using a config file stored in an S3 bucket.
+	clientset, err := authenticate.CreateClientFromS3Bucket(*bucket, *kubeconfig, *region, *ctx)
+	if err != nil {
+		log.Fatalln(err.Error())
```
Comment (jasonBirchall) [inline reports/live-1-domains/main.go:45]: Can you handle this error. For example, replace the above with `fmt.Errorf("Unable to authenticate to the cluster:", err)`

File: reports/live-1-domains/main.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	// Get all ingress resources
+	ingressList, err := ingress.GetAllIngressesFromCluster(clientset)
+	if err != nil {
+		log.Fatalln(err.Error())
```
Comment (jasonBirchall) [inline reports/live-1-domains/main.go:51]: Same as above, just add some context to why it failed.

File: reports/live-1-domains/main.go (hunk 6 above)
Comment (jasonBirchall) [reply reports/live-1-domains/main.go:45]: https://pkg.go.dev/fmt#example-Errorf

File: reports/live-1-domains/main.go (hunk 2 above)
Comment (jackstockley89) [reply reports/live-1-domains/main.go:55]: Yes that is fine i can you update

File: reports/live-1-domains/main.go (hunk 3 above)
Comment (jackstockley89) [reply reports/live-1-domains/main.go:73]: This is used by Live1DomainSearch. It creates and then IngressList that that is passed to Live1DomainSearch

File: reports/live-1-domains/main.go (hunk 7 above)
Comment (jackstockley89) [reply reports/live-1-domains/main.go:51]: Added a error message

File: reports/live-1-domains/main.go (hunk 6 above)
Comment (jackstockley89) [reply reports/live-1-domains/main.go:45]: Added a error message

File: reports/live-1-domains/main.go (hunk 4 above)
Comment (jackstockley89) [reply reports/live-1-domains/main.go]: have moved the function from public to internal

File: cloud-platform-reports-cronjobs/templates/live-one-domains.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+              valueFrom:
+                secretKeyRef:
+                  name: aws-creds
+                  key: secret-access-key
+            {{- include "cloud-platform-reports-cronjobs.kubeconfigLocation" . | indent 12 }}
+            {{- include "cloud-platform-reports-cronjobs.hoodaw-credentials" . | indent 12 }}
+            command:
+            - /bin/sh
+            - -c
+            - ./live-one-domains
```
[the remark this answers is not in the document]
Comment (jackstockley89) [reply cloud-platform-reports-cronjobs/templates/live-one-domains.yaml]: Yes this is incorrect updated the cronjob

File: cloud-platform-reports-cronjobs/templates/live-one-domains.yaml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+              valueFrom:
+                secretKeyRef:
+                  name: aws-creds
+                  key: secret-access-key
+            {{- include "cloud-platform-reports-cronjobs.kubeconfigLocation" . | indent 12 }}
+            {{- include "cloud-platform-reports-cronjobs.hoodaw-credentials" . | indent 12 }}
+            command:
+            - /bin/sh
+            - -c
+            - ./live-1-domains
```
Comment (poornima-krishnasamy) [inline cloud-platform-reports-cronjobs/templates/live-one-domains.yaml:37]: @jackstockley89 , this is incorrect. The module name for the report is `live1-domains` so the binary will be in that name and not `live-1-domains`

Comment (jasonBirchall) [reply cloud-platform-reports-cronjobs/templates/live-one-domains.yaml:37]: What was done about this?

File: reports/live-1-domains/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,19 @@
+FROM golang:1.17
```
Comment (jasonBirchall) [inline reports/live-1-domains/Dockerfile:1]: Can we use 1.18?

File: reports/live-1-domains/main_test.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+				},
+			},
+			want: []map[string]string{
+				{
+					"hostname":  "example.live-1.cloud-platform.service.justice.gov.uk",
+					"namespace": "namespace-1",
+					"ingress":   "ingress-1",
+				},
+			},
+			wantErr: false,
```
Comment (jasonBirchall) [inline reports/live-1-domains/main_test.go:50]: This is great. You have table to test the "good" path of this function. It's also handy to test the "bad" path, which will allow future devs to catch errors. The "bad" path in this instance would be a host name without "live-1" defined and the list returning empty.

File: reports/live-1-domains/main.go (hunk 5 above)
Comment (jasonBirchall) [reply reports/live-1-domains/main.go]: ignored

File: reports/live-1-domains/main.go (hunk 3 above)
Comment (jasonBirchall) [reply reports/live-1-domains/main.go]: function not used so it was removed.

File: reports/live-1-domains/Dockerfile (hunk 10 above)
Comment (jackstockley89) [reply reports/live-1-domains/Dockerfile:1]: yes should be no problem in changing to 1.18 I have been using this locally

**Review by jasonBirchall — APPROVED** [review]

**Pull request merged**